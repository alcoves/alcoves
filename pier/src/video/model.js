const mongoose = require('mongoose');
const shortid = require('shortid');
const Schema = mongoose.Schema;

const defaultThumbnail =
  'https://s3.us-east-2.wasabisys.com/media-bken/files/default-thumbnail-sm.jpg';

const fileSchema = new Schema({
  link: { type: String },
  startedAt: { type: Date },
  completedAt: { type: Date },
  status: { type: String, required: true },
  percentCompleted: { type: Number, default: 0, required: true },
});

const videoSchema = new Schema(
  {
    title: { type: String, required: true, default: shortid.generate },
    status: { type: String, required: true, default: 'uploading' },
    sourceFile: { type: String, required: false },
    _id: { type: String, default: shortid.generate },
    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },
    thumbnail: { type: String, default: defaultThumbnail, required: true },
    views: { type: Number, default: 0, required: true },
    files: {
      '480p': { type: fileSchema },
      '720p': { type: fileSchema },
      '1080p': { type: fileSchema },
      '1440p': { type: fileSchema },
      '2160p': { type: fileSchema },
      highQuality: { type: fileSchema },
    },
  },
  { timestamps: true }
);

videoSchema.index({ title: 'text' });

module.exports = mongoose.model('Video', videoSchema);
