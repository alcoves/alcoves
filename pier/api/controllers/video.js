const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

const Video = require('../models/video');

exports.upload = async (req, res) => {
  try {
    const user = await User.findOne({ email: req.body.email });
    if (user) {
      const passwordsMatch = await bcrypt.compare(
        req.body.password,
        user.password
      );

      if (passwordsMatch) {
        const accessToken = jwt.sign(
          { email: user.email, userId: user.id },
          process.env.JWT_KEY,
          { expiresIn: '1h' }
        );

        // TODO :: add hashed refreshToken to database

        const refreshToken = jwt.sign(
          { email: user.email, userId: user.id },
          process.env.JWT_KEY,
          { expiresIn: '1d' }
        );

        res
          .cookie('refreshToken', refreshToken, {
            httpOnly: true,
          })
          .status(200)
          .send({
            message: 'login succeeded',
            accessToken,
          });
      } else {
        res.status(401).send({ message: 'authentication failed' });
      }
    } else {
      res.status(401).send({ message: 'authentication failed' });
    }
  } catch (error) {
    throw error;
  }
};

exports.get = async (req, res) => {
  try {
    const userExists = await User.find({ email: req.body.email });

    if (userExists.length >= 1) {
      return res.status(409).send({
        message: 'user already exists',
      });
    }

    const user = new User({
      _id: mongoose.Types.ObjectId(),
      email: req.body.email,
      password: await bcrypt.hash(req.body.password, 10),
    });

    await user.save();
    res.status(201).send({ message: 'user created' });
  } catch (error) {
    throw error;
  }
};

exports.delete = async (req, res) => {
  try {
    const result = await User.deleteOne({ _id: req.params.userId });
    if (result.deletedCount >= 1) {
      res.status(200).send({ message: 'user deleted' });
    } else {
      res.status(400).send({ message: 'user was not deleted' });
    }
  } catch (error) {
    throw error;
  }
};
