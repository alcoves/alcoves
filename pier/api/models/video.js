const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const videoSchema = new Schema({
  _id: Schema.Types.ObjectId,
  title: { type: String, required: true },
  status: { type: String, required: false },
  sourceUrl: { type: String, required: false },
  sourceFileName: { type: String, required: true },
  author: { type: Schema.Types.ObjectId, ref: 'User', required: true },
});

module.exports = mongoose.model('Video', videoSchema);
