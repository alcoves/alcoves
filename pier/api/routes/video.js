const express = require('express');
const c = require('../controllers/video');

const router = express.Router();

router.get('/', c.get);
router.post('/login', c.upload);
router.delete('/:userId', c.delete);

module.exports = router;
