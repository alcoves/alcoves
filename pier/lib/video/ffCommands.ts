import { Metadata } from "./getMetadata"

// CalcMaxBitrate uses the videos original bitrate to determine what the max should be
// func CalcMaxBitrate(originalWidth int, desiredWidth int, bitrate int) int {
// vidRatio := float64(desiredWidth) / float64(originalWidth)
// return int(vidRatio * float64(bitrate) / 1000)
// }

function calcResizeFilter (w: number): string {
  return `scale=${w}:${w}:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2`
}

export function getX264Args (m: Metadata, width: number, url: string): string {
  let videoFilter = calcResizeFilter(width)
  if (m.video.r_frame_rate) {
    console.info("Applying framerate to video filter")
    videoFilter += `,fps=fps=${m.video.r_frame_rate}`
  }

  const commands = [
    "-y",
    "-i", url,
    "-c:v", "libx264",
    "-vf", videoFilter,
    "-crf", "22",
    "-preset", "medium",
    "-bf", "2",
    "-coder", "1",
    "-profile", "high",
    `./${width}p.mp4`
  ]

  if (m?.format?.bit_rate) {
    // TODO :: apply max bitrate
    // https://github.com/bkenio/tidal/blob/main/utils/x264.go
  } else {
    // TODO :: apply default bitrates
  }

  return commands.join(" ")
}