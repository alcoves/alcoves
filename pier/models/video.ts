import mongoose from 'mongoose'

export default mongoose.model("Video", new mongoose.Schema({
  _id: mongoose.Types.ObjectId,
  status: String,
  title: String,
  duration: Number,
  views: Number,
  pod: { type : mongoose.Types.ObjectId, ref: 'Pod' },
  owner: { type : mongoose.Types.ObjectId, ref: 'User' },
}, { timestamps: true }))