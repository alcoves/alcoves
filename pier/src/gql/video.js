const Video = require('../models/video');
const convertObjectToDotNotation = require('../lib/convertObjectToDotNotation');
const { gql } = require('apollo-server-express');

const typeDefs = gql`
  extend type Query {
    videos: [Video!]!
    videosByUserId(id: ID!): [Video!]!
    video(id: ID!): Video!
  }
  extend type Mutation {
    updateVideo(id: ID!, input: UpdateVideoInput!): Video!
    updateVideoFile(id: ID!, input: UpdateVideoFileInput!): Video!
  }
  type Video {
    id: ID!
    views: Int!
    title: String!
    user: User!
    status: String!
    thumbnail: String!
    createdAt: String!
    modifiedAt: String!
    sourceFile: String!
    files: [VideoFile!]
  }
  type VideoFile {
    link: String
    status: String!
    preset: String!
    createdAt: String
    modifiedAt: String
    percentCompleted: Float!
  }
  input UpdateVideoInput {
    title: String
    thumbnail: String
    status: String
  }
  input UpdateVideoFileInput {
    link: String
    preset: String
    status: String
    createdAt: String
    modifiedAt: String
    percentCompleted: Float
  }
`;

const resolvers = {
  Query: {
    video: async (_, { id }) => {
      return Video.findOne({ _id: id }).populate(
        'user',
        '_id avatar displayName'
      );
    },
    // Deprecate? could be used to get homepage video feed
    videos: async (_, input) => {
      return Video.find()
        .sort({ createdAt: -1 })
        .populate('user', '_id avatar displayName');
    },
    videosByUserId: async (_, { id }) => {
      return Video.find({ user: id })
        .sort({ createdAt: -1 })
        .populate('user', '_id avatar displayName followers');
    },
  },
  Mutation: {
    updateVideo: async (_, { id, input }, { user }) => {
      // if (!user) throw new Error('authentication failed');
      await Video.updateOne(
        { _id: id },
        { $set: convertObjectToDotNotation(input) }
      );

      return Video.findOne({ _id: id }).populate(
        'user',
        '_id avatar displayName'
      );
    },
    updateVideoFile: async (_, { id, input }, { user }) => {
      const video = await Video.findOne({ _id: id });
      let shouldUpdate;

      video.files.map(({ preset }) => {
        if (preset === input.preset) shouldUpdate = true;
      });

      if (shouldUpdate) {
        const convertedArrToDot = Object.entries(input).reduce(
          (acc, [k, v]) => {
            acc[`files.$.${k}`] = v;
            return acc;
          },
          {}
        );

        // console.log(`files.preset`, input.preset);
        // console.log('convertedArrToDot', convertedArrToDot);

        await Video.updateOne(
          { _id: id, 'files.preset': input.preset },
          { $set: convertedArrToDot }
        );
      } else {
        await Video.updateOne({ _id: id }, { $push: { files: input } });
      }

      return Video.findOne({ _id: id }).populate(
        'user',
        '_id avatar displayName'
      );
    },
  },
};

module.exports = { typeDefs, resolvers };
