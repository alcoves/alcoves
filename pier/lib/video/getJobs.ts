import { Metadata } from "./getMetadata"
import mongoose from "mongoose"
import { getX264Args} from './ffCommands'

export interface Preset {
  name: string
  width: number,
  height: number,
  defaultMaxRate: number
}

export function clampPreset (w: number, h: number, dw: number, dh: number): boolean {
  if ((w >= dw && h >= dh) || (w >= dh && h >= dw)) {
    return true
  }
  return false
}

export function getS3CpCommand(id: string) {
  return `aws s3 cp ./ s3://cdn.bken.io/v/${id}/progressive/ --recursive --profile wasabi --endpoint https://us-east-2.wasabisys.com`
}

export function getTranscodingJobs(m: Metadata, videoId: string, url: string) {
  const w = m.video.width || 0
  const h = m.video.height || 0

  const presets: unknown[] = [{
    _id: new mongoose.Types.ObjectId(),
    video: new mongoose.Types.ObjectId(videoId),
    type: 'transcode',
    status: 'queued',
    name: "transcode_360",
    ffmpegCommand: getX264Args(m, 480, 360, url),
  }]

  if (clampPreset(w, h, 1280, 720)) {
    presets.push({
      _id: new mongoose.Types.ObjectId(),
      video: new mongoose.Types.ObjectId(videoId),
      type: 'transcode',
      status: 'queued',
      name: "transcode_720",
      ffmpegCommand: getX264Args(m, 1280, 720, url),
    })
  
  }

  if (clampPreset(w, h, 1920, 1080)) {
    presets.push({
      _id: new mongoose.Types.ObjectId(),
      video: new mongoose.Types.ObjectId(videoId),
      type: 'transcode',
      status: 'queued',
      name: "transcode_1080",
      ffmpegCommand: getX264Args(m, 1920, 1080, url),
    })
  }

  if (clampPreset(w, h, 2560, 1440)) {
    presets.push({
      _id: new mongoose.Types.ObjectId(),
      video: new mongoose.Types.ObjectId(videoId),
      type: 'transcode',
      status: 'queued',
      name: "transcode_1440",
      ffmpegCommand: getX264Args(m, 2560, 1440, url),
    })
  }

  if (clampPreset(w, h, 3840, 2160)) {
    presets.push({
      _id: new mongoose.Types.ObjectId(),
      video: new mongoose.Types.ObjectId(videoId),
      type: 'transcode',
      status: 'queued',
      name: "transcode_2160",
      ffmpegCommand: getX264Args(m, 3840, 2160, url),
    })
  }

  return presets
}

export function getThumbnailJobs(m: Metadata, videoId: string, url: string) {
  const ffThumbCmds = [
    "-y",
    "-i", url,
    "-vf", "scale=854:480:force_original_aspect_ratio=increase,crop=854:480",
    "-vframes", "1",
    "-q:v", "6",
    "-f", "image2",
    "./thumb.jpg"
  ]
  const jobs: unknown[] = [{
    _id: new mongoose.Types.ObjectId(),
    video: new mongoose.Types.ObjectId(videoId),
    type: 'thumbnail',
    status: 'queued',
    name: "thumbnail_jpg",
    ffmpegCommand: ffThumbCmds.join(" ")
  }]
  return jobs
}

export function getPackagingJobs(videoId: string) {
  const jobs: unknown[] = [{
    _id: new mongoose.Types.ObjectId(),
    video: new mongoose.Types.ObjectId(videoId),
    type: 'package',
    status: 'queued',
    name: "package_hls",
  }]
  return jobs
}