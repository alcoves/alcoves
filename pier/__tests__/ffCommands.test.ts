import { x264, x264Package } from "../lib/video/ffCommands"
import { getPresets, Preset} from "../lib/video/getJobs"
import { getMetadata, Metadata } from "../lib/video/getMetadata"

describe("ffCommands", () => {
  test("that ffmpeg commands are generated 1/3", () => {
    const metadata: Metadata = {
      audio: {
        index: 0,
      },
      video: {
        index: 0,
        width: 2280,
        height: 1080
      },
      format: {
        width: 2280,
        height: 1080
      }
    }

    const presets: Preset[] = [
      {
        name: "360p",
        width: 1920,
        height: 1080,
        defaultMaxRate: 3000
      }
    ]

    expect(x264(metadata, presets)).toMatchSnapshot()
  })

  test("that ffmpeg commands are generated 2/3", () => {
    const metadata: Metadata = {
      audio: {
        index: 0,
      },
      video: {
        index: 0,
        width: 2280,
        height: 1080
      },
      format: {
        width: 2280,
        height: 1080
      }
    }

    const presets: Preset[] = [
      {
        name: "360p",
        width: 1920,
        height: 1080,
        defaultMaxRate: 3000
      }
    ]

    const x264Args = x264(metadata, presets)
    expect(x264Package(x264Args)).toMatchSnapshot()
  })
})

describe('full ffmpeg command generation tests', () => {
  test("that ffmpeg commands are generated 3/3", () => {
    const metadata: Metadata = {
      audio: {
        index: 0,
      },
      video: {
        index: 0,
        width: 1920,
        height: 1080
      },
      format: {
        width: 1920,
        height: 1080
      }
    }

    const presets = getPresets(metadata)
    const x264Args = x264(metadata, presets)
    const fullArgs = x264Package(x264Args)

    expect(presets).toMatchSnapshot()
    expect(x264Args).toMatchSnapshot()
    expect(fullArgs).toMatchSnapshot()
    expect(fullArgs.join(' ')).toMatchSnapshot()
  })
})