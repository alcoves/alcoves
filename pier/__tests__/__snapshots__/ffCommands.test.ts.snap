// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ffCommands that ffmpeg commands are generated 1`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:a?:0",
  "-c:a:1",
  "aac",
  "-b:a:0",
  "128k",
  "-ar:1",
  "48000",
  "-pix_fmt yuv420p",
  "-force_key_frames expr:gte(t,n_forced*2)",
  "-hls_playlist 1",
  "-hls_segment_type mpegts",
  "-hls_playlist_type vod",
  "-hls_time 2",
  "./stream.m3u8",
]
`;

exports[`ffCommands that ffmpeg commands are generated 1/3 1`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
]
`;

exports[`ffCommands that ffmpeg commands are generated 2`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:a?:0",
  "-c:a:1",
  "aac",
  "-b:a:0",
  "128k",
  "-ar:1",
  "48000",
  "-pix_fmt yuv420p",
  "-force_key_frames expr:gte(t,n_forced*2)",
  "-hls_playlist 1",
  "-hls_segment_type mpegts",
  "-hls_playlist_type vod",
  "-hls_time 2",
  "./stream.m3u8",
]
`;

exports[`ffCommands that ffmpeg commands are generated 2/3 1`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:a?:0",
  "-c:a:1",
  "aac",
  "-b:a:0",
  "128k",
  "-ar:1",
  "48000",
  "-pix_fmt yuv420p",
  "-force_key_frames expr:gte(t,n_forced*2)",
  "-hls_playlist 1",
  "-hls_segment_type mpegts",
  "-hls_playlist_type vod",
  "-hls_time 2",
  "./stream.m3u8",
]
`;

exports[`ffCommands that ffmpeg commands are generated 3/3 1`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:a?:0",
  "-c:a:1",
  "aac",
  "-b:a:0",
  "128k",
  "-ar:1",
  "48000",
  "-pix_fmt yuv420p",
  "-force_key_frames expr:gte(t,n_forced*2)",
  "-hls_playlist 1",
  "-hls_segment_type mpegts",
  "-hls_playlist_type vod",
  "-hls_time 2",
  "./stream.m3u8",
]
`;

exports[`full ffmpeg command generation tests that ffmpeg commands are generated 3/3 1`] = `
Array [
  Object {
    "defaultMaxRate": 1500,
    "height": 360,
    "name": "360",
    "width": 640,
  },
  Object {
    "defaultMaxRate": 8000,
    "height": 720,
    "name": "720",
    "width": 1280,
  },
  Object {
    "defaultMaxRate": 12000,
    "height": 1080,
    "name": "1080",
    "width": 1920,
  },
]
`;

exports[`full ffmpeg command generation tests that ffmpeg commands are generated 3/3 2`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=640:640:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:v:0",
  "-c:v:1",
  "libx264",
  "-filter:v:1",
  "scale=1280:1280:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:1",
  "high",
  "-map",
  "0:v:0",
  "-c:v:2",
  "libx264",
  "-filter:v:2",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:2",
  "high",
]
`;

exports[`full ffmpeg command generation tests that ffmpeg commands are generated 3/3 3`] = `
Array [
  "-map",
  "0:v:0",
  "-c:v:0",
  "libx264",
  "-filter:v:0",
  "scale=640:640:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:0",
  "high",
  "-map",
  "0:v:0",
  "-c:v:1",
  "libx264",
  "-filter:v:1",
  "scale=1280:1280:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:1",
  "high",
  "-map",
  "0:v:0",
  "-c:v:2",
  "libx264",
  "-filter:v:2",
  "scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2",
  "-crf",
  "22",
  "-preset",
  "medium",
  "-bf",
  "2",
  "-coder",
  "1",
  "-profile:v:2",
  "high",
  "-map",
  "0:a?:0",
  "-c:a:1",
  "aac",
  "-b:a:0",
  "128k",
  "-ar:1",
  "48000",
  "-pix_fmt yuv420p",
  "-force_key_frames expr:gte(t,n_forced*2)",
  "-hls_playlist 1",
  "-hls_segment_type mpegts",
  "-hls_playlist_type vod",
  "-hls_time 2",
  "./stream.m3u8",
]
`;

exports[`full ffmpeg command generation tests that ffmpeg commands are generated 3/3 4`] = `"-map 0:v:0 -c:v:0 libx264 -filter:v:0 scale=640:640:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2 -crf 22 -preset medium -bf 2 -coder 1 -profile:v:0 high -map 0:v:0 -c:v:1 libx264 -filter:v:1 scale=1280:1280:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2 -crf 22 -preset medium -bf 2 -coder 1 -profile:v:1 high -map 0:v:0 -c:v:2 libx264 -filter:v:2 scale=1920:1920:force_original_aspect_ratio=decrease,scale=trunc(iw/2)*2:trunc(ih/2)*2 -crf 22 -preset medium -bf 2 -coder 1 -profile:v:2 high -map 0:a?:0 -c:a:1 aac -b:a:0 128k -ar:1 48000 -pix_fmt yuv420p -force_key_frames expr:gte(t,n_forced*2) -hls_playlist 1 -hls_segment_type mpegts -hls_playlist_type vod -hls_time 2 ./stream.m3u8"`;
